### ISSUE TITLE
recommend registering client-issued CIDs with the proxy first

### ISSUE THREAD

**Issue by marten-seemann** [ietf-wg-masque/draft-ietf-masque-quic-proxy#84]
When the client wishes to issue a new CID for the target, it might be advantageous to register it with the proxy first (and wait for acknowledgement), before actually sending it to the target. That way, if the proxy rejects the CID, there’s no risk that the target already started using that CID, which would lead to the loss of all packets sent using that CID, until that CID is retired (which takes at least 1 RTT).

**Comment by marten-seemann** [comment]
Looks like I opened this issue too early (reading through the document from top to bottom).

This is already recommended in section 4.2. However, in section 4 it says:
> The client sends a REGISTER_CLIENT_CID capsule whenever it advertises a new Client Connection ID to the target

This sentence could probably be reworded by replacing “whenever” with “before”.

**Comment by tfpauly** [comment]
Yes, it would be useful to clarify the ordering — saying "before" is good there, or at least in the same packet to the proxy so the proxy is able to read the content first before passing along a packet to the next hop.

**Comment by tfpauly** [comment]
As a note here, this text:

> If the rejection indicated a conflict due to the client CID, the client MUST select a new Connection ID before sending a new request, and generate a new packet. For example, if a client is sending a QUIC Initial packet and chooses a Connection ID that conflicts with an existing mapping to the same target server, it will need to generate a new QUIC Initial.[¶](https://ietf-wg-masque.github.io/draft-ietf-masque-quic-proxy/draft-ietf-masque-quic-proxy.html#section-4.9.1-5)

Explains that a client needs to resend the initial, but doesn't describe how the proxy drops this initial, or how it knows which packets to allow/drop.

**Comment by tfpauly** [comment]
Option:
- If the proxy sends a close client CID capsule, it will drop all packets sent by the client until the proxy sends an acknowledgement of a new client CID. This regresses us into a slower case when we have a conflict. The client can be told to only send the initial flight before getting the ack or close, and to then not send anything more until it gets an ack.

**Comment by tfpauly** [comment]
Fixed by #121
